### ISSUE TITLE
Train layoutlmv3 with custom dataset  by loading from local directory 

### ISSUE THREAD

**Issue by deepanshudashora** [NielsRogge/Transformers-Tutorials#123]
How to train layoutlmv3 with custom dataset  by loading from local directory,

**Comment by jyotiyadav94** [comment]
@deepanshudashora 

where u able to create your own dataset ? you can try using Label Studio OCR.

**Comment by photopea** [comment]
Hi @NielsRogge , is it possible to get in touch with you somehow? I could not find your email address, your twitter messages are blocked. If you won a billion dollars, there would be no way to tell you about it :D

I am Ivan from www.Photopea.com , wihch is probably the best free photo editor that exists today :) Would you be interested in some kind of a cooperation, that is ? We have millions of users. We would like to add AI features, which could run on the Hugging Face infrastructure. You can write me on Twitter: https://twitter.com/photopeacom  or support@photopea.com

**Comment by jyotiyadav94** [comment]
Hi @NielsRogge 

I am getting the below error. I am using the code 
![Screenshot 2022-06-17 183257](https://user-images.githubusercontent.com/72126242/174412644-b97c750e-1e63-46a1-944e-0bb6c1dbd67b.png)
```
I am using the below code `import os
import numpy as np
import pandas as pd
import torch
from transformers import LayoutLMv2Tokenizer, LayoutLMv2ForTokenClassification, LayoutLMv2Config
from torchvision.transforms import ToTensor
from torch.utils.data import Dataset, DataLoader
from PIL import ImageDraw, ImageFont, Image
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
import warnings 
warnings.filterwarnings('ignore')

def normalize_box(box, width, height):
    width = int(width)
    height = int(height)
    # [ <class 'str'> 1901 <class 'int'>
    print("YYYY: ", box,'*',type(box[0]),'*',  width,'*', type(width) ,'*')
    return [
         int(1000 * (int(box[0]) / width)),
         int(1000 * (int(box[1]) / height)),
         int(1000 * (int(box[2]) / width)),
         int(1000 * (int(box[3]) / height)),
     ]

def de_normalize(box, width, height):
  return [
          int((width * box[0])/1000),
          int((height * box[1])/1000),
          int((width * box[2])/1000),
          int((height * box[3])/1000)
  ]

def resize_and_align_bounding_box(bbox, original_image, target_size):
    x_, y_ = original_image.size
    x_scale = target_size / x_ 
    y_scale = target_size / y_
    origLeft, origTop, origRight, origBottom = tuple(bbox)
    x = int(np.round(origLeft * x_scale))
    y = int(np.round(origTop * y_scale))
    xmax = int(np.round(origRight * x_scale))
    ymax = int(np.round(origBottom * y_scale)) 
    return [x-0.5, y-0.5, xmax+0.5, ymax+0.5]
 
class InvoiceDataSet(Dataset):
    """LayoutLM dataset with visual features."""
 
    def __init__(self, df, tokenizer, max_length, target_size, train=True):
        self.df = df
        self.tokenizer = tokenizer
        self.max_seq_length = max_length
        self.target_size = target_size
        self.pad_token_box = [0, 0, 0, 0]
        self.train = train
 
    def __len__(self):
        return self.df.shape[0]
    
    def __getitem__(self, idx):
        item = self.df.iloc[idx,:].to_dict()        
        #base_path = data_config.base_image_path
        original_image = Image.open(os.path.join('/content/data/' , item["image_path"])).convert("RGB")
        # resize to target size (to be provided to the pre-trained backbone)
        resized_image = original_image.resize((self.target_size, self.target_size))
        # first, read in annotations at word-level (words, bounding boxes, labels)
        words = item["words"]
        # print(words)
        # print("IN GETITEM XXX: ", type(words), type(words[0]))
        unnormalized_word_boxes = item["bbox"]
        # print(unnormalized_word_boxes)

        # print("IN boxes XXX: ", type(unnormalized_word_boxes), type(unnormalized_word_boxes[0]))
        word_labels = item["label"]
        # print(word_labels)

        width = item["imageWidth"]
        height = item["imageHeight"]
        normalized_word_boxes = [normalize_box(bbox, width, height) for bbox in unnormalized_word_boxes]
        # print('normalized_word_boxes',normalized_word_boxes)
        assert len(words) == len(normalized_word_boxes)
 
        # next, transform to token-level (input_ids, attention_mask, token_type_ids, bbox, labels)
        token_boxes = []
        unnormalized_token_boxes = []
        token_labels = []
        for word, unnormalized_box, box, label in zip(words, unnormalized_word_boxes, normalized_word_boxes, word_labels):
            word_tokens = self.tokenizer.tokenize(word)
            unnormalized_token_boxes.extend(unnormalized_box for _ in range(len(word_tokens)))
            token_boxes.extend(box for _ in range(len(word_tokens)))
            # label first token as B-label (beginning), label all remaining tokens as I-label (inside)
            for i in range(len(word_tokens)):
                if  1 == 1:#i == 0:
                    token_labels.extend(['B-' + label])
                else:
                    token_labels.extend(['I-' + label])
        
        # Truncation of token_boxes + token_labels
        special_tokens_count = 2 
        if len(token_boxes) > self.max_seq_length - special_tokens_count:
            token_boxes = token_boxes[: (self.max_seq_length - special_tokens_count)]
            unnormalized_token_boxes = unnormalized_token_boxes[: (self.max_seq_length - special_tokens_count)]
            token_labels = token_labels[: (self.max_seq_length - special_tokens_count)]
        
        # add bounding boxes and labels of cls + sep tokens
        token_boxes = [self.pad_token_box] + token_boxes + [[1000, 1000, 1000, 1000]]
        unnormalized_token_boxes = [self.pad_token_box] + unnormalized_token_boxes + [[1000, 1000, 1000, 1000]]
        token_labels = [-100] + token_labels + [-100]
        print('token_labels',token_labels)
        # return token_labels

        encoding = self.tokenizer(' '.join(words), padding='max_length', truncation=True)
        # Padding of token_boxes up the bounding boxes to the sequence length.
        print('encoding',encoding)
        input_ids = self.tokenizer(' '.join(words), truncation=True)["input_ids"]
        padding_length = self.max_seq_length - len(input_ids)
        token_boxes += [self.pad_token_box] * padding_length
        unnormalized_token_boxes += [self.pad_token_box] * padding_length
        token_labels += [-100] * padding_length
        encoding['bbox'] = token_boxes
        encoding['labels'] = token_labels
 
        assert len(encoding['input_ids']) == self.max_seq_length
        assert len(encoding['attention_mask']) == self.max_seq_length
        assert len(encoding['token_type_ids']) == self.max_seq_length
        assert len(encoding['bbox']) == self.max_seq_length
        assert len(encoding['labels']) == self.max_seq_length
 
        encoding['resized_image'] = ToTensor()(resized_image)
        # rescale and align the bounding boxes to match the resized image size (typically 224x224) 
        encoding['resized_and_aligned_bounding_boxes'] = [resize_and_align_bounding_box(bbox, original_image, self.target_size) for bbox in unnormalized_token_boxes]
        #encoding['unnormalized_token_boxes'] = unnormalized_token_boxes
        
        # finally, convert everything to PyTorch tensors 
        for k,v in encoding.items():
            if k == 'labels':
                label_indices = []
                # convert labels from string to indices
                for label in encoding[k]:
                    if label != -100:
                        label_indices.append(data_config.label2id[label])
                    else:
                        label_indices.append(label)
                encoding[k] = label_indices
            encoding[k] = torch.as_tensor(encoding[k])
        print(encoding)
        return encoding
 ```

**Comment by aditya11ad** [comment]
hi I created the annotations folder containing **json** files ,  like this: 
```
{
    "form": [
        {
            "box": [
                84,
                109,
                136,
                119
            ],
            "text": "23456789",
            "label": "invoice_num",
            "words": [
                {
                    "box": [
                        84,
                        109,
                        136,
                        119
                    ],
                    "text": "23456789"
                }
            ]
.
.
.

```

please guide me, how can I train layoutlmV3 on this.

**Comment by NielsRogge** [comment]
Hi,

You don't necessarily have to write a script like the one for FUNSD. You can just create a custom PyTorch dataset, which I explain here: https://github.com/NielsRogge/Transformers-Tutorials/issues/123#issuecomment-1156256273

**Comment by NielsRogge** [comment]
Hi @photopea, I've forwarded your request to the team, someone will reach out :)

**Comment by aditya11ad** [comment]
> 

hi,
what would be the structure of dataframe that contains the words + boxes for each document image.
if any sample example df is available plz provide.

**Comment by aditya11ad** [comment]
hi 
i have prepared the dataframe like this:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/20943198/183844502-d77c5aca-b883-43d5-a2f4-b4dc103a7971.png">

now what should be the scrip for fine tuning ?

**Comment by pavel-nesterov** [comment]
> now what should be the scrip for fine tuning ?
Hi @aditya11ad  
THis might be helpful https://github.com/ruifcruz/sroie-on-layoutlm/blob/main/LayoutLM_fine_tunning_for_SROIE_dataset.ipynb

**Comment by PoonamS25** [comment]
Hi,
Can we train layoutLM on custom dataset with .txt annotations files (yolo format annotations files) available on local machine?

**Comment by DANISHFAYAZNAJAR** [comment]
> Hi,
> 
> You can create a regular PyTorch Dataset as follows:
> 
> ```
> from torch.utils.data import Dataset
> from PIL import Image
> 
> class CustomDataset(Dataset):
>      def __init__(self, root, df, processor):
>           self.root = root
>           self.df = df
>           self.processor = processor
> 
>      def __getitem__(self, idx):
>           # get document image + corresponding words and boxes
>           item = self.df.iloc[idx]
>           image = Image.open(self.root + ...).convert('RGB')
>           words = item.words
>           boxes = item.boxes
> 
>           # use processor to prepare everything for the model
>           encoding = self.processor(image, words, boxes=boxes)
> 
>          return encoding
> ```
> 
> This is just a draft, assuming you have a root folder with all your document images, and a Pandas dataframe that contains the words + boxes for each document image.
> 
> You can then instantiate the dataset as follows:
> 
> ```
> from transformers import LayoutLMv3Processor
> 
> processor = LayoutLMv3Processor.from_pretrained("microsoft/layoutlmv3-base")
> 
> dataset = CustomDataset(root="path_to_your_root", df="your_dataframe", processor=processor)
> ```

@NielsRogge  What about the Labels. Using OCR we can get words and bounding boxes. But you  haven't mentioned anything about labels in it. I believe we need to generate somehow labels also and use them in it. Can you clarify whether we need those or not.

**Comment by vcjayan** [comment]
@NielsRogge If I do LayoutLM training with custom invoice images, what should be the annotation format ? Should I use Q-A format like FUNSD (invoice_number_Q & invoice_number_A, date_Q & date_A etc.) or can I just annotate all labels directly like invoice_number, invoice_date, vendor_name etc ?

**Comment by NielsRogge** [comment]
@DANISHFAYAZNAJAR if you have labels at the word level (like the FUNSD dataset has), then you can do the following:
```
# get document image + corresponding words and boxes
item = self.df.iloc[idx]
image = Image.open(self.root + ...).convert('RGB')
words = item.words
boxes = item.boxes
word_labels = item.ner_tags

# use processor to prepare everything for the model
encoding = self.processor(image, words, boxes=boxes, word_labels=word_labels, return_tensors="pt")

# remove batch dimension which the processor adds by default
encoding = {k:v.squeeze() for k,v in encoding.items()}

return encoding
```

**Comment by NielsRogge** [comment]
@vcjayan you just need a list of words, their boxes and their labels for each document. 

So this could look like:
``` 
words = ["hello", "world", "this", "is", "invoice", "number", '14721"]
boxes = [[1,2,3,4] for _ in range(len(words))]
word_labels = ["other", "other", "other", "other", "other", "other", "invoice_number"]
```
assuming you have 2 classes ("other" and "invoice_number")

**Comment by ankitarajsharma** [comment]
@NielsRogge How would we specify the train and test splits. I am using this - https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLMv3/Fine_tune_LayoutLMv3_on_FUNSD_(HuggingFace_Trainer).ipynb notebook. I tried the above provided code the dataset is returned as 
dataset
<__main__.CustomDataset at 0x7a1c979eead0> object whereas the expectation was a DatasetDict object - What am I missing? Please help

**Comment by NielsRogge** [comment]
You can create 2 instances, like so:

```
train_dataset = CustomDataset(dataset=dataset["train"], processor=processor)
val_dataset = CustomDataset(dataset=dataset["validation"], processor=processor)
```

**Comment by madhavi1102** [comment]
> Hi,
> 
> You can create a regular PyTorch Dataset as follows:
> 
> ```
> from torch.utils.data import Dataset
> from PIL import Image
> 
> class CustomDataset(Dataset):
>      def __init__(self, root, df, processor):
>           self.root = root
>           self.df = df
>           self.processor = processor
> 
>      def __getitem__(self, idx):
>           # get document image + corresponding words and boxes
>           item = self.df.iloc[idx]
>           image = Image.open(self.root + ...).convert('RGB')
>           words = item.words
>           boxes = item.boxes
> 
>           # use processor to prepare everything for the model
>           encoding = self.processor(image, words, boxes=boxes)
> 
>          return encoding
> ```
> 
> This is just a draft, assuming you have a root folder with all your document images, and a Pandas dataframe that contains the words + boxes for each document image.
> 
> You can then instantiate the dataset as follows:
> 
> ```
> from transformers import LayoutLMv3Processor
> 
> processor = LayoutLMv3Processor.from_pretrained("microsoft/layoutlmv3-base")
> 
> dataset = CustomDataset(root="path_to_your_root", df="your_dataframe", processor=processor)
> ```

@NielsRogge for this CustomDataset, does the return item in __getitem__ method is only encoding or we can separately return 
 return {
            "input_ids": torch.tensor(self.encoding["input_ids"][index], dtype=torch.int64),
            "attention_mask": torch.tensor(self.encoding["attention_mask"][index], dtype=torch.int64),
            "bbox": torch.tensor(self.encoding["bbox"], dtype=torch.int64),
            "pixel_values": torch.tensor(self.encoding['pixel_values'], dtype=torch.float32),
            "labels": torch.tensor(self.encoding['labels'], dtype=torch.int64)
        }

**Comment by Dhananjay-97** [comment]
> hi I created the annotations folder containing **json** files , like this:
> 
> ```
> {
>     "form": [
>         {
>             "box": [
>                 84,
>                 109,
>                 136,
>                 119
>             ],
>             "text": "23456789",
>             "label": "invoice_num",
>             "words": [
>                 {
>                     "box": [
>                         84,
>                         109,
>                         136,
>                         119
>                     ],
>                     "text": "23456789"
>                 }
>             ]
> .
> .
> .
> ```
> 
> please guide me, how can I train layoutlmV3 on this.

PLease tell me how did you generate the annotations?

**Comment by Aesthethic0de** [comment]
hi i trained on form dataset annotated documents with 200 images
model trained for 100 epochs
but i am not getting any result while inferencing

**Comment by vidya-chandran** [comment]
@NielsRogge Can you help me on creating a mapping between the predicted labels and tokens(words) associated. I tried extracting text from bounding box which was not accurate. can we create a direct mapping from the labels to the words. Many thanks in advance

**Comment by yashakagf** [comment]
> hi I created the annotations folder containing **json** files , like this:
> 
> ```
> {
>     "form": [
>         {
>             "box": [
>                 84,
>                 109,
>                 136,
>                 119
>             ],
>             "text": "23456789",
>             "label": "invoice_num",
>             "words": [
>                 {
>                     "box": [
>                         84,
>                         109,
>                         136,
>                         119
>                     ],
>                     "text": "23456789"
>                 }
>             ]
> .
> .
> .
> ```
> 
> please guide me, how can I train layoutlmV3 on this.

Hey can you guide me , how you prepared dataset in this format

**Comment by dsoft-jvo** [comment]
> You can create 2 instances, like so:
> 
> ```
> train_dataset = CustomDataset(dataset=dataset["train"], processor=processor)
> val_dataset = CustomDataset(dataset=dataset["validation"], processor=processor)
> ```

@NielsRogge How should the CustomDataset class be edited to allow this function?
